src/sources/gcp_pubsub.rs

```diff
@@ -214,24 +245,59 @@ impl SourceConfig for PubsubConfig {
 
         let endpoint = self.endpoint.as_deref().unwrap_or(PUBSUB_URL).to_string();
         let uri: Uri = endpoint.parse().context(UriSnafu)?;
+
+        let tls = TlsSettings::from_options(&self.tls)?;
+        let host = uri.host().unwrap_or("pubsub.googleapis.com");
+        let mut tls_config = ClientTlsConfig::new().domain_name(host);
+        if let Some((cert, key)) = tls.identity_pem() {
+            tls_config = tls_config.identity(Identity::from_pem(cert, key));
+        }
+        for authority in tls.authorities_pem() {
+            tls_config = tls_config.ca_certificate(Certificate::from_pem(authority));
+        }
+
+        let mut endpoint = Channel::from_shared(endpoint).context(EndpointSnafu)?;
+        if uri.scheme() != Some(&Scheme::HTTP) {
+            endpoint = endpoint.tls_config(tls_config).context(EndpointTlsSnafu)?;
+        }
+
+        let token_generator = match &credentials {
+            Some(credentials) => credentials.clone().spawn_regenerate_token(),
+            None => {
+                let (sender, receiver) = watch::channel(());
+                // Dropping the sender will cause the receiver to
+                // error, but there is no good way of hanging on to it
+                // and sending nothing. This `forget` will leak one
+                // sender per instance of this source and so never
+                // drop it.
+                std::mem::forget(sender);
+                receiver
+            }
```

Comment:
It's going to leak at most a few hundred bytes every config (re)load. It would have to run for many days with hundreds of reloads a day before it grew larger than a rounding error.

The alternative would be spawning off another task just to hold it open waiting for all the receivers to close. I can do that if desired but it's not obvious that would actually be less resource usage.

Reply:
I'd be for spawning a task to await the receiver going away.

In terms of raw resource usage, it would certainly consume more memory, but I'm hesitant to leak anything more than we absolutely have to.

Reply:
Fixed in fff1663